### PR TITLE
Feature/subscribe feature change

### PR DESCRIPTION
**Pull request by 0xchimz**
- update apollo store with `_id`
- create subscribe for feature change
- add loading and disable to submit button

### CODE REVIEW

File: src/server/graphql/MutationRootType.js [hunk 1]
```diff
@@ -32,12 +32,16 @@ const MutationRootType = new GraphQLObjectType({
...
-      resolve: async (_, { _id }) => Collections.update('feature', {
-        _id: Mongodb.ObjectId(_id)
-      }, { $set: { active: false } })
+      resolve: async (_, { _id }) => {
+        const updatedFeature = await Collections.update('feature', {
+          _id: Mongodb.ObjectId(_id)
+        }, { $set: { active: false } })
+        pubsub.publish('featureChanged', updatedFeature.value)
+        return updatedFeature.value
+      }
```
Comment (pNagi) [inline src/server/graphql/MutationRootType.js]:
```js
const featureUpdated = await Collections.update(...code...).value
pubsub.publish('featureUpdated', featureUpdated)
return featureUpdated
```

or

```js
const result = await Collections.update(...code...)
pubsub.publish('featureUpdated', result.value)
return result.value
```

File: src/server/graphql/MutationRootType.js [hunk 2]
```diff
@@ -32,12 +32,16 @@ const MutationRootType = new GraphQLObjectType({
...
       description: 'Close a feature',
       args: { _id: { type: GraphQLString } },
-      resolve: async (_, { _id }) => Collections.update('feature', {
-        _id: Mongodb.ObjectId(_id)
-      }, { $set: { active: false } })
+      resolve: async (_, { _id }) => {
+        const updatedFeature = await Collections.update('feature', {
+          _id: Mongodb.ObjectId(_id)
+        }, { $set: { active: false } })
+        pubsub.publish('featureChanged', updatedFeature.value)
```
Comment (pNagi) [inline src/server/graphql/MutationRootType.js]:
```js
pubsub.publish('featureUpdated', featureUpdated)
```

or

```js
pubsub.publish('featureUpdated', result.value)
```

File: src/server/graphql/SubscriptionRootType.js [hunk 3]
```diff
@@ -9,6 +9,11 @@ const SubscriptionRootType = new GraphQLObjectType({
       type: FeatureType,
       args: { productId: { type: GraphQLString } },
       resolve: (_) => _
+    },
+    featureChanged: {
```
Comment (pNagi) [inline src/server/graphql/SubscriptionRootType.js]: `featureUpdated`

File: src/server/mongodb/commands/Update.js [hunk 4]
```diff
@@ -1,4 +1,4 @@
-const FindItems = db => async (collectionName, query, update) =>
-  db.collection(collectionName).update(query, update)
+const UpdateItem = db => async (collectionName, query, update) =>
+  db.collection(collectionName).findOneAndUpdate(query, update, { returnOriginal: false })
```
Comment (pNagi) [inline src/server/mongodb/commands/Update.js]: I think this line is too long.

```js
 db.collection(collectionName).findOneAndUpdate(query, update, {
  returnOriginal: false
})
```

File: src/shared/nt-features/CreateFeatureForm.react.js [hunk 5]
```diff
@@ -32,11 +40,7 @@ class CreateFeatureForm extends React.Component {
 
   constructor (props) {
     super(props)
-    this.state = {
-      name: undefined,
-      A: 50,
-      B: 50,
-    }
+    this.state = Object.assign({}, initialState)
```
Comment (pNagi) [inline src/shared/nt-features/CreateFeatureForm.react.js]: I don't like `Object.assign()`

how about `{ ...initialState }`

File: src/shared/nt-features/CreateFeatureForm.react.js [hunk 6]
```diff
@@ -65,19 +69,26 @@ class CreateFeatureForm extends React.Component {
   }
 
   handleSubmit = e => {
+    e.preventDefault()
```
Comment (pNagi) [inline src/shared/nt-features/CreateFeatureForm.react.js]: Why `e.preventDefault()` before mutate`? (Really don't know if the order matter)

File: src/shared/nt-features/CreateFeatureForm.react.js [hunk 7]
```diff
@@ -65,19 +69,26 @@ class CreateFeatureForm extends React.Component {
...
+              B: this.state.B,
+            }
           }
-        },
-      })
-    } else {
-      e.preventDefault()
+        })
+        .then(() => {
+          this.setState(Object.assign({}, initialState))
```
Comment (pNagi) [inline src/shared/nt-features/CreateFeatureForm.react.js]: Same as above comment.

File: src/shared/nt-features/CreateFeatureForm.react.js [hunk 8]
```diff
@@ -112,7 +124,15 @@ class CreateFeatureForm extends React.Component {
           {this.renderProportionInput('B')}
         </div>
         <div className={styles.nt__submit}>
-          <input type='submit' value='Submit' />
+          <button
+            type='submit'
+            className={
+              classNames({ [styles['--loading']]: this.state.loading })
+            }
```
Comment (pNagi) [inline src/shared/nt-features/CreateFeatureForm.react.js]:
```js
className={classNames({
  [styles['--loading']]: this.state.loading
})}
```

File: src/shared/nt-features/CreateFeatureForm.styl [hunk 9]
```diff
@@ -43,17 +54,35 @@
...
+        top 'calc(50% - (%s / 2))' % $l-size
+        left 'calc(50% - (%s / 2))' % $l-size
+        display block
+        width $l-size
+        height $l-size
+        border $l-border solid c-gray50
+        border-top $l-border solid c-gray500
+        border-radius 50%
+        content ''
+        animation spinAround $l-spin-time infinite linear
```
Comment (pNagi) [inline src/shared/nt-features/CreateFeatureForm.styl]: duplicate css

File: src/shared/nt-features/FeaturesPage.react.js [hunk 10]
```diff
@@ -28,6 +29,15 @@ const FEATURES_SUBSCRIPTION = gql`
   }
 `
 
+const FEATURE_CHANGED_SUBSCRIPTION = gql`
```
Comment (pNagi) [inline src/shared/nt-features/FeaturesPage.react.js]: `FEATURE_UPDATED_SUBSCRIPTION`

File: src/shared/nt-features/FeaturesPage.react.js [hunk 11]
```diff
@@ -28,6 +29,15 @@ const FEATURES_SUBSCRIPTION = gql`
...
+  subscription onFeatureChanged ($productId: String!) {
```
Comment (pNagi) [inline src/shared/nt-features/FeaturesPage.react.js]: `onFeatureUpdated`

File: src/shared/nt-features/FeaturesPage.react.js [hunk 12]
```diff
@@ -28,6 +29,15 @@ const FEATURES_SUBSCRIPTION = gql`
...
+    featureChanged (productId: $productId) {
```
Comment (pNagi) [inline src/shared/nt-features/FeaturesPage.react.js]: `featureUpdated`

File: src/shared/nt-features/FeaturesPage.react.js [hunk 13]
```diff
@@ -73,13 +83,13 @@ class FeaturesPage extends React.Component {
...
       loading: PropTypes.bool,
-      subscribeToMore: PropTypes.func,
+      subscribeToMore: PropTypes.func
     }).isRequired
   }
 
   componentDidMount () {
     this.props.data.subscribeToMore({
-      document: FEATURES_SUBSCRIPTION,
+      document: FEATURE_ADDED_SUBSCRIPTION,
```
Comment (pNagi) [inline src/shared/nt-features/FeaturesPage.react.js]: `FEATURE_UPDATED_SUBSCRIPTION`

File: src/shared/nt-features/FeaturesPage.react.js [hunk 14]
```diff
@@ -89,6 +99,18 @@ class FeaturesPage extends React.Component {
...
+    this.props.data.subscribeToMore({
+      document: FEATURE_CHANGED_SUBSCRIPTION,
+      variables: { productId: this.props.productId },
+      updateQuery: (prev, { subscriptionData }) => {
+        if (!subscriptionData.data) return prev
+
+        const updatedFeature = subscriptionData.data.featureChanged
+
+        const index = findIndex(prev.features, { _id: updatedFeature._id })
+        return update(prev, { features: { [index]: { active: { $set: updatedFeature.active } } } })
```
Comment (pNagi) [inline src/shared/nt-features/FeaturesPage.react.js]: I think this line is too long.

**Review by pNagi — CHANGES_REQUESTED** [review]

File: src/shared/nt-features/CreateFeatureForm.react.js (hunk 6 above)
Comment (0xchimz) [reply src/shared/nt-features/CreateFeatureForm.react.js]: order is matter. if there is sync function before `e.preventDefault()`, events will run a code before canceling them.

File: src/shared/nt-features/CreateFeatureForm.react.js [hunk 15]
```diff
@@ -4,6 +4,7 @@ import React from 'react'
 import { graphql } from 'react-apollo'
 import { compose } from 'recompose'
 
+import LoadingComponent from '../nt-uikit/Loading.react'
```
Comment (pNagi) [inline src/shared/nt-features/CreateFeatureForm.react.js]:
```js
import { Loading } from '../nt-uikit'
```

File: src/shared/nt-features/CreateFeatureForm.react.js [hunk 16]
```diff
@@ -112,7 +124,16 @@ class CreateFeatureForm extends React.Component {
           {this.renderProportionInput('B')}
         </div>
         <div className={styles.nt__submit}>
-          <input type='submit' value='Submit' />
+          <button
+            type='submit'
+            disabled={this.state.loading}
+          >
+            <LoadingComponent
```
Comment (pNagi) [inline src/shared/nt-features/CreateFeatureForm.react.js]: `<Loading`

File: src/shared/nt-features/FeaturesPage.react.js [hunk 17]
```diff
@@ -91,6 +100,24 @@ class FeaturesPage extends React.Component {
         return update(prev, { features: { $unshift: [ newFeature ] } })
       }
     })
+    this.props.data.subscribeToMore({
+      document: FEATURE_UPDATED_SUBSCRIPTION,
+      variables: { productId: this.props.productId },
+      updateQuery: (prev, { subscriptionData }) => {
+        if (!subscriptionData.data) return prev
+
+        const updatedFeature = subscriptionData.data.featureChanged
```
Comment (pNagi) [inline src/shared/nt-features/FeaturesPage.react.js]: `featureUpdated`

File: webpack.config.prod.babel.js [hunk 18]
```diff
@@ -8,10 +8,19 @@ import UglifyJSPlugin from 'uglifyjs-webpack-plugin'
...
 export default {
   bail: true,
-  entry: [
-    path.join(__dirname, 'src/shared/nt-styles/base.styl'),
-    path.join(__dirname, 'src/client/client.prod.js'),
-  ],
+  // entry: [
+  //   path.join(__dirname, 'src/shared/nt-styles/base.styl'),
+  //   path.join(__dirname, 'src/client/client.prod.js'),
+  // ],
```
Comment (pNagi) [inline webpack.config.prod.babel.js]: please remove comments

**Review by pNagi — CHANGES_REQUESTED** [review]

**Review by pNagi — APPROVED** [review]

**Pull request merged**